dace/codegen/tools/type_inference.py

```diff
@@ -0,0 +1,580 @@
+    _dispatch(t.body, symbols, inferred_symbols)
+
+
+def _For(t, symbols, inferred_symbols):
+    _generic_For(t, symbols, inferred_symbols)
+
+
+def _AsyncFor(t, symbols, inferred_symbols):
+    _generic_For(t, symbols, inferred_symbols)
+
+
+def _If(t, symbols, inferred_symbols):
+    _dispatch(t.test, symbols, inferred_symbols)
+    _dispatch(t.body, symbols, inferred_symbols)
+
+    while (t.orelse and len(t.orelse) == 1
+           and isinstance(t.orelse[0], ast.If)):
+        t = t.orelse[0]
+        _dispatch(t.test, symbols, inferred_symbols)
+        _dispatch(t.body, symbols, inferred_symbols)
+
+    # final else
+    if t.orelse:
+        _dispatch(t.orelse, symbols, inferred_symbols)
+
+
+def _While(t, symbols, inferred_symbols):
+    _dispatch(t.test, symbols, inferred_symbols)
+    _dispatch(t.body, symbols, inferred_symbols)
+
+
+def _Str(t, symbols, inferred_symbols):
+    return dtypes.pointer(dtypes.int8)
+
+
+def _FormattedValue(t, symbols, inferred_symbols):
+    # FormattedValue(expr value, int? conversion, expr? format_spec)
+    _dispatch(t.value, symbols, inferred_symbols)
+
+    if t.format_spec is not None:
+        if not isinstance(t.format_spec, ast.Str):
+            _dispatch(t.format_spec, symbols, inferred_symbols)
+
+
+def _JoinedStr(t, symbols, inferred_symbols):
+    for value in t.values:
+        if not isinstance(value, ast.Str):
+            _dispatch(value, symbols, inferred_symbols)
+    return dtypes.pointer(dtypes.int8)
+
+
+def _Name(t, symbols, inferred_symbols):
+    if t.id in cppunparse._py2c_reserved:
+        return dtypes.typeclass(np.result_type(t.id))
+    else:
+        # check if this name is a python type, it is in defined_symbols or in local symbols.
+        # If yes, take the type
+        inferred_type = None
+        if t.id.strip("()") in cppunparse._py2c_typeconversion:
+            inferred_type = cppunparse._py2c_typeconversion[t.id.strip("()")]
+        elif t.id in symbols:
+            # defined symbols could have dtypes, in case convert it to typeclass
+            inferred_type = symbols[t.id]
+            if isinstance(inferred_type, np.dtype):
+                inferred_type = dtypes.typeclass(inferred_type.type)
+        elif t.id in inferred_symbols:
+            inferred_type = inferred_symbols[t.id]
+        return inferred_type
+
+
+def _NameConstant(t, symbols, inferred_symbols):
+    return dtypes.typeclass(np.result_type(t.value).name)
+
+
+def _Num(t, symbols, inferred_symbols):
+    # get the minimum between the minimum type needed to represent this number and the corresponding default data types
+    # e.g., if num=1, then it will be represented by using the default integer type (int32 if C data types are used)
+    return dtypes.result_type_of(
+        dtypes.typeclass(type(t.n)),
+        dtypes.typeclass(np.min_scalar_type(t.n).name))
+    np.promote_types(np.min_scalar_type(t.n), dtypes.typeclass(type(t.n)).type)
```

Comment:
Is something crazy happening with indentation here, or is this last statement never reached?